### PR TITLE
Differentiable PolyFEM

### PR DESCRIPTION
**Pull request by zfergus**
Differentiable PolyFEM using the adjoint method.

Credit: @arvigj @Huangzizhou

### CODE REVIEW

File: input-spec.json [hunk 1]
```diff
@@ -1303,20 +1303,12 @@
         "min": 0,
         "doc": "Number of time steps"
     },
-    {
```
Comment (teseoch) [inline input-spec.json:1306]: revert

File: .github/workflows/continuous.yml [hunk 2]
```diff
@@ -153,7 +149,7 @@ jobs:
   ####################
 
   Windows:
-    runs-on: windows-2022
+    runs-on: [self-hosted,Windows]
```
Comment (teseoch) [inline .github/workflows/continuous.yml]: revert

File: input-spec.json [hunk 3]
```diff
@@ -1464,6 +1456,7 @@
         "optional": [
             "enable_overwrite_solver",
             "solver",
+            "adjoint_solver",
```
Comment (teseoch) [inline input-spec.json:1459]: write an issue

File: src/polyfem/State.hpp [hunk 4]
```diff
@@ -561,6 +576,64 @@ namespace polyfem
...
+		{
+			const int actual_dim = problem->is_scalar() ? 1 : mesh->dimension();
+			if (mixed_assembler == nullptr)
+				return actual_dim * n_bases;
+			else
+				return actual_dim * n_bases + n_pressure_bases;
+		}
+
+		// Aux functions for setting up adjoint equations
+		void compute_force_hessian(const Eigen::MatrixXd &sol, const Eigen::MatrixXd &disp_grad, StiffnessMatrix &hessian);
```
Comment (zfergus) [inline src/polyfem/State.hpp:600]: Should this be named force_jacobian?

File: src/polyfem/assembler/CMakeLists.txt [hunk 5]
```diff
@@ -48,10 +48,14 @@ set(SOURCES
 	RhsAssembler.hpp
 	SaintVenantElasticity.cpp
 	SaintVenantElasticity.hpp
+	ViscousDamping.cpp
```
Comment (teseoch) [inline src/polyfem/assembler/CMakeLists.txt]: duplicated

File: src/polyfem/assembler/CMakeLists.txt [hunk 6]
```diff
@@ -48,10 +48,14 @@ set(SOURCES
...
+	ViscousDamping.hpp
```
Comment (zfergus) [inline src/polyfem/assembler/CMakeLists.txt:52]: remove me

File: src/polyfem/main_optimization.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
+#include <filesystem>
```
Comment (teseoch) [inline src/polyfem/main_optimization.cpp:1]: delete this file

File: src/polyfem/mesh/mesh2D/Mesh2D.cpp [hunk 8]
```diff
@@ -98,6 +98,26 @@ namespace polyfem
 			igl::barycentric_coordinates(p, A, B, C, coords);
 		}
 
+		void Mesh2D::compute_face_jacobian(const int el_id, const Eigen::MatrixXd &reference_map, Eigen::MatrixXd &jacobian) const
```
Comment (teseoch) [inline src/polyfem/mesh/mesh2D/Mesh2D.cpp:101]: create an issue

File: src/polyfem/solver/DenseNewtonDescentSolver.tpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
+#pragma once
```
Comment (teseoch) [inline src/polyfem/solver/DenseNewtonDescentSolver.tpp:1]: open issue to merge both newtons

File: src/polyfem/solver/FullNLProblem.hpp [hunk 10]
```diff
@@ -26,6 +26,9 @@ namespace polyfem::solver
 		virtual void gradient(const TVector &x, TVector &gradv) override;
 		virtual void hessian(const TVector &x, THessian &hessian);
 
+		virtual TVector component_values(const TVector &x);
```
Comment (teseoch) [inline src/polyfem/solver/FullNLProblem.hpp]: remove

File: src/polyfem/solver/forms/BodyForm.cpp [hunk 11]
```diff
@@ -80,4 +110,236 @@ namespace polyfem::solver
...
+	{
+		const auto &bases = rhs_assembler_.bases();
+		const auto &gbases = rhs_assembler_.gbases();
+		const int dim = rhs_assembler_.mesh().dimension();
+		const int actual_dim = rhs_assembler_.problem().is_scalar() ? 1 : dim;
+
+		const int n_elements = int(bases.size());
+		term.setZero(n_verts * dim, 1);
+
+		auto storage = utils::create_thread_storage(LocalThreadVecStorage(term.size()));
```
Comment (teseoch) [inline src/polyfem/solver/forms/BodyForm.cpp:124]: refactor somewhere else

File: src/polyfem/solver/forms/ElasticForm.hpp [hunk 12]
```diff
@@ -52,6 +52,19 @@ namespace polyfem::solver
 		/// @param x Current solution at time t
 		void update_quantities(const double t, const Eigen::VectorXd &x) override { x_prev_ = x; }
 
+		/// @brief Compute the derivative of the force wrt lame/damping parameters, then multiply the resulting matrix with adjoint_sol.
+		/// @param[in] x Current solution
+		/// @param[in] adjoint Current adjoint solution
+		/// @param[out] term Derivative of force multiplied by the adjoint
+		void foce_material_derivative(const Eigen::MatrixXd &x, const Eigen::MatrixXd &x_prev, const Eigen::MatrixXd &adjoint, Eigen::VectorXd &term);
```
Comment (zfergus) [inline src/polyfem/solver/forms/ElasticForm.hpp:59]: is this a typo? `foce`

File: src/polyfem/solver/forms/ElasticForm.cpp [hunk 13]
```diff
@@ -84,4 +109,230 @@ namespace polyfem::solver
...
+	{
+		const int dim = is_volume_ ? 3 : 2;
+
+		const int n_elements = int(bases_.size());
+
+		if (assembler_.name() == "ViscousDamping")
+		{
+			term.setZero(2);
+
+			auto storage = utils::create_thread_storage(LocalThreadVecStorage(term.size()));
```
Comment (teseoch) [inline src/polyfem/solver/forms/ElasticForm.cpp:123]: refactor elsewhere

File: src/polyfem/state/StateInit.cpp [hunk 14]
```diff
@@ -343,6 +372,29 @@ namespace polyfem
...
 			problem->set_parameters(args["preset_problem"]);
 		}
+
+		if (optimization_enabled)
+		{
+			if (is_contact_enabled())
+			{
+				if (!args["contact"]["use_convergent_formulation"])
+				{
+					args["contact"]["use_convergent_formulation"] = true;
```
Comment (teseoch) [inline src/polyfem/state/StateInit.cpp:382]: open issue

File: src/polyfem/utils/MatrixUtils.cpp [hunk 15]
```diff
@@ -383,6 +383,17 @@ Eigen::MatrixXd polyfem::utils::unflatten(const Eigen::VectorXd &x, int dim)
 	return X;
 }
 
+void polyfem::utils::vector2matrix(const Eigen::VectorXd &vec, Eigen::MatrixXd &mat)
+{
+	int size = sqrt(vec.size());
```
Comment (zfergus) [inline src/polyfem/utils/MatrixUtils.cpp:388]: do we already have an unflatten?

File: src/polyfem/utils/CMakeLists.txt [hunk 16]
```diff
@@ -38,12 +40,14 @@ set(SOURCES
 	RBFInterpolation.hpp
 	RefElementSampler.cpp
 	RefElementSampler.hpp
+	CubicHermiteSplineParametrization.hpp
```
Comment (teseoch) [inline src/polyfem/utils/CMakeLists.txt:43]: remove

File: src/polyfem/utils/CMakeLists.txt [hunk 17]
```diff
@@ -38,12 +40,14 @@ set(SOURCES
...
 	Selection.cpp
 	Selection.hpp
 	StringUtils.cpp
 	StringUtils.hpp
 	Timer.hpp
 	Types.hpp
+	PolarDecomposition.hpp
```
Comment (teseoch) [inline src/polyfem/utils/CMakeLists.txt:50]: remove

File: src/polyfem/utils/CMakeLists.txt [hunk 18]
```diff
@@ -22,6 +22,8 @@ set(SOURCES
 	InterpolatedFunction.hpp
 	Interpolation.cpp
 	Interpolation.hpp
+	IsosurfaceInflator.cpp
```
Comment (teseoch) [inline src/polyfem/utils/CMakeLists.txt:25]: remove

**Review by teseoch — APPROVED** [review]

**Pull request merged**